test/parse.test.js

```diff
@@ -105,3 +105,43 @@ test('parse long', function (t) {
   t.throws(pb(1.5), null, 'float throws');
   t.end();
 });
+
+test('parse buffer', function (t) {
+  t.ok(parse(0));
+  t.equal(parse(Buffer.from([ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1 ])).toString(), '127.0.0.1');
```

Comment:
No... this is wrong.  ::127.0.0.1 is NOT an IPv4 mapped address, it's literally ::127.0.0.1 or ::7f00:0.  You have other mistakes of this form.

Reply:
My intention was to try to keep the same kind of functionality the existing code has.

E.g. `ip6addr.parse('127.0.0.1').toBuffer()` returns `<Buffer 00 00 00 00 00 00 00 00 00 00 ff ff c0 a8 00 01>`. So `ip6addr.parse(Buffer.from(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 255, 255, 127, 0, 0, 1)).toString()` should return the orginal `'127.0.0.1'`.

It also seems that `ip6addr.parse(Buffer.from(0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 127, 0, 0, 1)).toString()` should return `'127.0.0.1'`.

Perhaps if the original `ip6addr.parse('127.0.0.1').toBuffer()` returned a 4-byte buffer `<Buffer c0 a8 00 01>`, then the buffer-parsing function could be modified to distinguish between true ipv4 and mapped ipv4 addresses, but I didn't want to modify any of the existing functionality.

I'm not sure there's a path forward for this PR, but thanks for your work in considering it!

Reply:
I reexamined the internal data structure for `parse('127.0.0.1')`, `parse('::ffff:127.0.0.1') and `parse('::ffff:7f00:01')`.

They are like this:

**ip.parse('127.0.0.1')**
```
Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ipv4Bare: true, ipv4Mapped: true } }
```

**ip.parse('::ffff:127.0.0.1')**
```
Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ipv4Mapped: true } }
```

**ip.parse('::ffff:7f00:1')**
```
Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: {} }
```

A buffer-parsing function could avoid setting any `_attrs`. I.e., there might not be a good reason to treat `Buffer<00 00 00 00 00 00 00 00 00 00 ff ff, 7f 00 00 01>` as a mapped ipv4 address as opposed to a plain ipv6 address.

Would that be preferred?

Then `parse(parse('::ffff:7f00:1').toBuffer()).toString()` would equal `'::ffff:7f00:1'`, but `parse(parse('127.0.0.1').toBuffer()).toString()` would also equal `'::ffff:7f00:1'`.

Reply:
There is a difference between '127.0.0.1' and '::127.0.0.1'.  The first is IPv4, which gets converted into '::ffff:127.0.0.1'.  The second is native IPv6 (albeit not used) and should not be converted to any sort of IPv4, whether mapped ('::ffff:<v4>') or native IPv4.

Reply:
To elaborate, the buffer data in line 111 should result in '::127.0.0.1' not '127.0.0.1', aka. '::ffff:127.0.0.1'.

Reply:
I agree. `Buffer<00 00 00 00 00 00 00 00 00 00 00 00 7f 00 00 01>` should not turn into '127.0.0.1' and should stay '::127.0.01' (or rather '::f700:1', based the other discussion). The way forward seems to be to not modify _field bytes 10 and 11 and to not set any _attr keys.

---

For Buffers like `Buffer<00 00 00 00 00 00 00 00 00 00 ff ff 7f 00 00 01>`, we seem to have a few options though:

The `Addr` created by `parse('127.0.0.1')`, `parse('::ffff:127.0.0.1')`, and `parse('::ffff:f700:1')` will all give this buffer when `toBuffer` is called. So which type of `Addr` should be created when parsing such a buffer?

1. `Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ip4Bare: true, ipv4Mapped: true } }` (what is created by `parse('127.0.0.1')`)
2. `Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { ipv4Mapped: true } }` (what is created by `parse('::ffff:127.0.0.1')`)
3. `Addr { _fields: [ 0, 0, 0, 0, 0, 65535, 32512, 1 ], _attrs: { } }` (what is created by `parse('::ffff:7f00:1')`)

My inclination is 3, i.e., not setting any keys of _attrs, but I think any of these options is reasonable.

Reply:
Your question is an interesting one, and I need to know "Where are you getting these Buffers from"? In an IPv6 packet you'll never see v4mapped, but some protocols may use IPv6 address space to express both.  Looking at what you feed via parse(), I'd go with option 2.

My concern this whole time has been you're reading '00 00 00 00 00 00 00 00 00 00 00 00 7f 00 00 01' incorrectly. '00 00 00 00 00 00 00 00 00 00 ff ff aaa bbb ccc ddd' should always become an IPv4 mapped address.